### ISSUE TITLE
[SPIKE] One approach for kubelet configuration

### ISSUE THREAD

**Issue by atsikham** [epiphany-platform/epiphany#2704]
**Is your spike related to a problem or idea? Please describe.**
At the moment we edit kubelet's config.yaml directly to specify values [[1](https://github.com/epiphany-platform/epiphany/blob/b19ae4dbb0ddb52a1bbb00646f462045de79798e/ansible/playbooks/roles/kubernetes_master/defaults/main.yml#L15)] [[2](https://github.com/epiphany-platform/epiphany/blob/b19ae4dbb0ddb52a1bbb00646f462045de79798e/ansible/playbooks/roles/kubernetes_node/defaults/main.yml#L7)] there. But there is [KubeletConfiguration](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) K8s object for that. This spike is to identify their relationship and decide which way we update kubelet's configuration.

**Describe the outcome you'd like**
Dive into the topic, choose the aproach and create a new task for implementation if it's not done in the scope of SPIKE.

**What is the reason or source for the spike**
`kubelet-config-<version>` is created by default by kubeadm with some default values. One day these values might overwrite our settings set directly in kubelet's config.yaml.

**Additional context**

https://github.com/epiphany-platform/epiphany/blob/b19ae4dbb0ddb52a1bbb00646f462045de79798e/ansible/playbooks/roles/kubernetes_common/tasks/configure-kubelet.yml#L21

---

**DoD checklist**

* [ ] Reader is able to understand the results of spike
* [ ] The results of the spike are presented in a table (to show simply what are compared or researched parameters) / not applicable
* [ ] Each value / cell in the results table is described more deeply below
* [ ] Demo of the spike (automated as much as possible)
* [ ] Design doc updated

**Comment by rafzei** [comment]
The `/var/lib/kubelet/config.yaml` already has `apiVersion: kubelet-config.v1beta1`
The file is actually uploaded a configMap in the cluster.

:arrow_right:  Current approach:
1. Variables are hardcoded and stored in the defaults:
```
kubelet_custom_config:
  evictionHard:
    memory.available: 200Mi
  kubeReserved:
    cpu: 50m
    memory: 256Mi
  systemReserved:
    cpu: 50m
    memory: 768Mi
```
2. If kubelet_custom_config is provided (de facto always) then `config.yaml.j2` will be used.
The config.yml.j2:
```
{{ kubelet_config_yaml.content | b64decode | from_yaml |
   combine(kubelet_custom_config, recursive=True) | to_nice_yaml(indent=2) }}
```
3. After the combination, the kubelet config.yml will contain values from point 1. If the values already exist, they will be overwritten.


:arrow_right: Opcional approach
1. Keep **all** fields in defaults or give user the possibility to provide them
2. Instead of `config.yml.j2`  would be `kubelet-config.v1beta1.yaml.j2` file
3. Instead of combing yamls, just use it in ansible template module

Basically, there is no significant change, the current approach is fine, we do not have to maintain the whole configuration file.

---
See how Kubespray uses that:
node:  
https://github.com/kubernetes-sigs/kubespray/blob/master/roles/kubernetes/node/templates/kubelet-config.v1beta1.yaml.j2
master:   
https://github.com/kubernetes-sigs/kubespray/blob/master/roles/kubernetes/control-plane/templates/kubeadm-config.v1beta2.yaml.j2
Versioning:  
https://github.com/kubernetes-sigs/kubespray/blob/4550f8c50f1e79f1ac49335af86279df09a662a6/roles/kubernetes/node/tasks/kubelet.yml#L11-L16

**Comment by atsikham** [comment]
In fact we can manage kubelet config file directly taking values from `kubelet_custom_config ` and via kubeadm [config](https://github.com/epiphany-platform/epiphany/blob/9fd3e99b50bf73920480798b8bd336688f3746a3/ansible/playbooks/roles/kubernetes_master/templates/kubeadm-config.yml.j2#L49) if they have no conflicts. @rafzei What do you think about managing those values only by [kubeadm config](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/) in a single place? Is it possible to define all values that we specify there?

**Comment by rafzei** [comment]
@atsikham Thanks for the comment.
 Looks like it is. According to documentation we could:
1- use one place for storing custom parameters 
2- use one place for generating kubelet configuration
3- use one kubelet configuration for all kubeadm activities like kubeadm init/join and even kubeadm upgrade 

Currently the parameters are hardcoded in [kubeadm config](https://github.com/epiphany-platform/epiphany/blob/9fd3e99b50bf73920480798b8bd336688f3746a3/ansible/playbooks/roles/kubernetes_master/templates/kubeadm-config.yml.j2)
On the node side, we use the defaults + merge I've mentioned above [config.yml.j2](https://github.com/epiphany-platform/epiphany/blob/9fd3e99b50bf73920480798b8bd336688f3746a3/ansible/playbooks/roles/kubernetes_common/templates/config.yaml.j2)

**Comment by seriva** [comment]
Not going todo anymore spikes for Epiphany.
